File: countValley.js
const path = [ "U" , "D" ,"D" , "D" , "U" , "D" , "U" , "U"];
const steps = 8;



const result = countValley(steps , path);
console.log(result);

function countValley(steps , path) {
    let countUpDown = 0;
    let valley = 0;
    
    for(let step of path) {
        if(step === "D") {
            countUpDown -= 1;
        }else if(step === "U") {
            countUpDown += 1;
            
            if(countUpDown === 0) {
                valley += 1;
            }
        }        
      
    }
    
    return valley;

}

File: src/CountValley.java

public class CountValley {
    public static void main(String[] args) {
        String paths = "DDUUUUDD";

        int valleys = 0;
        int countUpDown = 0;
        int steps = paths.length();
        char[] ch = paths.toCharArray();
        for(int step = 0; step < steps; step++) {
            if((ch[step]) == 'U'){
                countUpDown += 1;
                if(countUpDown == 0) {
                    valleys++;
                }
            }else {
                countUpDown -= 1;
            }
        }

        System.out.println(valleys);
    }
}